aimon/reprompting_api/pipeline.py

```diff
@@ -0,0 +1,544 @@
+        reprompter (Reprompter): Utility for generating corrective prompts based on evaluation feedback.
+        telemetry (TelemetryLogger): Logger for capturing telemetry data.
+        detect (Detect): AIMon detection client for evaluating model responses.
+        
+    Returns:
+        dict:
+            {
+                "best_response" (str): Best model response across all iterations.
+                "telemetry" (list, optional): Iteration-level telemetry if enabled.
+                "summary" (str, optional): Human-readable run summary if enabled.
+            }
+    """
+    def __init__(self, llm_fn, config):
+        """
+        Initialize pipeline with LLM callable and RepromptingConfig.
+        
+        Args:
+            llm_fn (callable): Function to call the LLM. 
+                Signature: llm_fn(recommended_prompt_template: Template, system_prompt: str, context: str, user_query: str) -> str
+            config (RepromptingConfig): Configuration object with API keys and limits.
+
+        """
+        self.llm_fn = llm_fn
+        self.config = config or RepromptingConfig()
+        
+        # Utilities for reprompting, telemetry, and scoring
+        self.reprompter = Reprompter()
+        self.telemetry = TelemetryLogger()
+
+        # Initialize AIMon Detect for response evaluation
+        self.detect = Detect(
+            values_returned=['user_query', 'instructions', 'generated_text', 'context'],
+            config={
+                "instruction_adherence": {
+                    "detector_name": "default",
+                    "explain": True,
+                    "extract_from_system": False
+                },
+                "groundedness": {
+                    "detector_name": "default",
+                    "explain": True
+                },
+                "toxicity": {
+                    "detector_name": "default",
+                    "explain": True
+                }
+            },
+            api_key=self.config.aimon_api_key,
+            application_name = self.config.application_name,
+            model_name = self.config.model_name,
+            publish=self.config.publish
+        )
+        
+    def run(self, system_prompt: str, context: str, user_query: str, user_instructions):
+        """
+        Execute the full re-prompting pipeline.
+        
+        Process:
+          1. Build an initial prompt with query, context, and instructions.
+          2. Call the LLM to generate a response.
+          3. Evaluate the response with AIMon detectors.
+          4. If violations are found, iteratively generate corrective prompts and re-prompt the LLM.
+          5. Stop when all instructions are followed or iteration limits are reached.
+          6. Return the best response (lowest residual error) along with telemetry and a summary if configured.
+        
+        Args:
+            user_query (str): The user's query or instruction.
```

Comment:
if there are constraints  like these fields required to be non-empty strings or non empty lists on these fields: user_query, content and user_instructions, it would be great to specify that